src/hotspot/share/services/memReporter.cpp

```diff
@@ -39,10 +39,28 @@ size_t MemReporterBase::committed_total(const MallocMemory* malloc, const Virtua
   return malloc->malloc_size() + malloc->arena_size() + vm->committed();
 }
 
-void MemReporterBase::print_total(size_t reserved, size_t committed) const {
+// There can be upto two CDS archives which can contain readonly data. On Windows, pages are not
+// shareable so the RO region may not actually be read only
+size_t MemReporterBase::readonly_total(FileMapInfo* info) const {
+  size_t total = 0;
+  if (info!= NULL) {
+    FileMapRegion* r = info->region_at(MetaspaceShared::ro);
+    if (r->read_only()) {
+      total = r->used();
+    }
+  }
+  return total;
+}
+
+void MemReporterBase::print_total(size_t reserved, size_t committed, size_t read_only) const {
```

Comment:
Since we now only print readonly for one tag in one report, we can keep this patch much smaller. The lLess complex the better. print_total gets called from several places that have nothing to do with CDS readonly.

My suggestion would be to CDS-specific stuff directly in `MemSummaryReporter::report_summary_of_type` since only there we only want the "Shared class" line in the summary report to show readonly. My proposal would be:

```
--- a/src/hotspot/share/services/memReporter.cpp
+++ b/src/hotspot/share/services/memReporter.cpp
@@ -214,7 +214,11 @@ void MemSummaryReporter::report_summary_of_type(MEMFLAGS flag,
     outputStream* out   = output();
     const char*   scale = current_scale();
     out->print("-%26s (", NMTUtil::flag_to_name(flag));
-    flag == mtClassShared ? print_total(reserved_amount, committed_amount, read_only_bytes) : print_total(reserved_amount, committed_amount);
+    print_total(reserved_amount, committed_amount);
+    if (flag == mtClassShared) {
+      output()->print(", readonly=" SIZE_FORMAT "%s",
+                      amount_in_current_scale(read_only_bytes), scale);
+    }
```